plugin/cache/handler.go

```diff
@@ -39,39 +39,28 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 		ttl = i.ttl(now)
 	}
 	if i == nil {
-		if !do {
-			setDo(rc)
-		}
 		crr := &ResponseWriter{ResponseWriter: w, Cache: c, state: state, server: server, do: do}
-		return plugin.NextOrFailure(c.Name(), c.Next, ctx, crr, rc)
+		return c.doRefresh(ctx, state, crr, do)
```

Comment:
this bool `do` should also be removed

Reply:
Why should it be removed? it's used to later set do if not already set.
Maybe we can always run this code regardless we are doing a fetch?

Reply:
that should already be in state or said differently the message in state should either have DO set or we should check and set it, no reason to carry a boolean argument here.

Reply:
ok, removed